algorithm/include/gnuradio-4.0/algorithm/SchmittTrigger.hpp

```diff
@@ -0,0 +1,259 @@
+#ifndef SCHMITTTRIGGER_HPP
+#define SCHMITTTRIGGER_HPP
+
+#include <cmath>
+#include <cstddef>
+
+#include <gnuradio-4.0/HistoryBuffer.hpp>
+#include <gnuradio-4.0/meta/UncertainValue.hpp>
+
+namespace gr::trigger {
+
+/**
+ * @brief Enumeration for the interpolation methods used for sub-sample
+ * precision.
+ */
+enum class InterpolationMethod {
+    NO_INTERPOLATION           = 0,
+    BASIC_LINEAR_INTERPOLATION = 1, /// basic linear interpolation
+    LINEAR_INTERPOLATION       = 2, /// interpolation via linear regression over multiple samples
+    POLYNOMIAL_INTERPOLATION   = 3  /// Savitzky–Golay filter-based methods
+};
+
+enum class EdgeDetection { NONE = 0, RISING = 1, FALLING = 2 };
+
+/**
+ * @brief A real-time capable digital Schmitt trigger implementation.
+ *
+ * @see https://en.wikipedia.org/wiki/Schmitt_trigger
+ * This class processes input samples and detects rising and falling edges based on specified thresholds.
+ * It supports sub-sample precision through various interpolation methods:
+ *  * NO_INTERPOLATION: nomen est omen
+ *  * BASIC_LINEAR_INTERPOLATION: basic linear interpolation based on the new and previous sample
+ *  * LINEAR_INTERPOLATION: interpolation via linear regression over the samples between when
+ *    the lower and upper threshold has been crossed and vice versa
+ *  * POLYNOMIAL_INTERPOLATION: Savitzky–Golay filter-based methods
+ *    https://en.wikipedia.org/wiki/Savitzky%E2%80%93Golay_filter
+ *     (TODO: WIP needs Tensor<T> and SVD implementation)
+ */
+template<typename T, InterpolationMethod Method = InterpolationMethod::NO_INTERPOLATION, std::size_t interpolationWindow = 16UZ>
+requires(std::is_arithmetic_v<T> or (UncertainValueLike<T> && std::is_arithmetic_v<meta::fundamental_base_value_type_t<T>>))
+struct SchmittTrigger {
+    using value_t  = meta::fundamental_base_value_type_t<T>;
+    using EdgeType = std::conditional_t<std::is_floating_point_v<value_t>, T, float>; // float being used for integer types
+
+    value_t _threshold{1};
+    value_t _offset{0};
+    value_t _upperThreshold;
+    value_t _lowerThreshold;
+
+    bool _lastState{false}; // true if above upper threshold
+
+    HistoryBuffer<T, interpolationWindow> _historyBuffer;
+
+    // edge timing variables
+    EdgeDetection lastEdge = EdgeDetection::NONE;
+    EdgeType      lastEdgeIdx{1};    /// relative index [-EdgeType|_MAX,1[ w.r.t. current sample of the last detected edge
+    EdgeType      lastEdgeOffset{0}; /// relative sub-sample offset [0, 1[ w.r.t. current sample of the last detected edge
```

Comment:
The thing is that -- normally it's:
 * `float` -> `float`
 * `double` -> `double`
 * `UncertainValue<float>` -> `UncertainValue<float>`
 * `UncertainValue<double>` -> `UncertainValue<double>`
 
 The integer types are a bit special because while the vertical values are discrete, the time-offsets aren't, thus my choice to default to `float` or `UncertainValue<float>` for them. Was also temporarily contemplating keeping them all as `float` or `UncertainValue<float>` but this would complicate the downstream logic probably.
 
 My goal was first to have a viable working solution and to see how it works out in the field.